force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,191 @@
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return Donation History list.
+    */
+    public List<DonationHistoryController.DonationHistory> getDonationHistory(Id contactId){
+        if(hasAccessToDonationHistory()){
+            List<OpportunityContactRole> credits = new DonationHistorySelector().getOpportunityContactRoleList(contactId);
+            List<DonationHistoryController.DonationHistory> result = buildDonationHistoryFromOCRs(credits);
+            return result;
+        } else {
+            throw new UTIL_Permissions.InsufficientPermissionException(Label.commonInsufficientPermissions);
+        }
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify each condition to list Donation.
+    */
+    private static List<DonationHistoryController.DonationHistory> buildDonationHistoryFromOCRs(List<OpportunityContactRole> credits){
+        List<DonationHistoryController.DonationHistory> donationHistory = new List<DonationHistoryController.DonationHistory>();
+        List<DonationHistoryController.DonationHistory> hardCredits = buildHardCreditHistoryFromOCRs(credits);
+        List<DonationHistoryController.DonationHistory> softCredits = buildSoftCreditHistoryFromOCRs(credits);
+        List<Id> softCreditIds = filterVerifiedIds(softCredits);
+        List<DonationHistoryController.DonationHistory> partialSoftCredits = verifyPartialSoftCreditCondition(softCreditIds, credits);
+        List<DonationHistoryController.DonationHistory> partialAndSoftCredits = buildPartialAndSoftCreditFromOCRs(softCredits, partialSoftCredits);
+        
+        donationHistory.addAll(hardCredits);
+        donationHistory.addAll(partialAndSoftCredits);
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> buildHardCreditHistoryFromOCRs(List<OpportunityContactRole> credits){
+        List<DonationHistoryController.DonationHistory> hardCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: credits){
+            if (ocr.isPrimary) {
+                DonationHistoryController.DonationHistory hardCredit = new DonationHistoryController.DonationHistory(ocr, HARD_CREDIT);
+                hardCreditList.add(hardCredit);
+            }
+        }
+        return hardCreditList;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> buildSoftCreditHistoryFromOCRs(List<OpportunityContactRole> donations){
+        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group 
+        Filter_Rule__mdt roles = CMT_FilterRule_SEL.getFilterRuleByDeveloperName(NPSP_CREDIT_ROLES);
```

Comment:
Another naming suggestion here, since the object is a Filter Rule we can name it after the type that it is.
```suggestion
        Filter_Rule__mdt filterRule = CMT_FilterRule_SEL.getFilterRuleByDeveloperName(NPSP_CREDIT_ROLES);
```

Reply:
changed